### PR TITLE
Support OidcProviderClient injection and token revocation

### LINKED ISSUE #44157: Make `quarkus-oidc`'s `OidcProviderClient` available to the custom code
### Description

There have been a couple of requests from users related to the current session's tokens having to be revoked, for example, when the user logs out, I believe Apple requires it in some contexts, and it may be required in other deployments. 

Right now, users are on their own on how to do it, for simple cases, an HTTP client form call might do, but it can be messy and should be avoidable.

CC @sschellh @FroMage 

### Implementation ideas

Add `OidcProviderClient#revokeToken`, make `OidcProviderClient` a CDI bean

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #44157.

@sschellh asked about an option for the user code be able to revoke tokens (I guess, either access or refresh tokens or both) when one of the Logout events is observed.
It reminded me that in Renarde, for the Apple authentication, this is also a typical requirement.

So this PR:
* makes `OidcProviderClient` which is used by `quarkus-oidc`, both injectable and also accessible as a `SecurityIdentity` attribute
* `SecurityEvent` listeners can react to logout, and other events like authentication failure, by using `OidcProviderClient` to revoke tokens if they need to - test is provided
* Or, the user code can use an `injected` `OidcProviderClient` to revoke tokens or do additional actions like token introspection. For example, Renarde code can have `@Inject OidcSession session;` , logout, and then follow up with revoking tokens.

The only remaining thing that I believe should be done is that `OidcProviderClient` should become an API interface, with the runtime subpackage providing an implementation

### CODE REVIEW

**Review by michalvavrik — COMMENTED** [review]
This looks good, but my understanding was that you want to move towards unification of the `io.quarkus.oidc.client.runtime.OidcClientImpl` and `io.quarkus.oidc.runtime.OidcProviderClient` as part of the https://github.com/quarkusio/quarkus/issues/39298. Wouldn't it be smarter to define some common API shared by both clients that you will expose? Otherwise when that happens, you will need to take into considerations users may already sue the `OidcProviderClient` when making changes.

**Comment by FroMage** [comment]
> This looks good, but my understanding was that you want to move towards unification of the `io.quarkus.oidc.client.runtime.OidcClientImpl` and `io.quarkus.oidc.runtime.OidcProviderClient` as part of the #39298. Wouldn't it be smarter to define some common API shared by both clients that you will expose? Otherwise when that happens, you will need to take into considerations users may already use the `OidcProviderClient` when making changes.

I didn't realise we have two client classes. Good point about this being confusing.

Also, this PR doesn't have any docs, which is what I'm most interested in :-/

**Comment by sberyozkin** [comment]
Hi @FroMage, `quarkus-oidc-client` is designed to be used independently of `quarkus-oidc`, primarily for acquiring tokens  when no human user is involved, for service applications.

What @michalvavrik pointed out is indeed a target, expand the scope of  what `quarkus-oidc-client` can do, for example, @stianst has ideas about using it in tools, for it be able to do more than just acquiring tokens, but also introspecting them to get their active status, or get some info about the user...

Michal, I actually started thinking about it during the `oidc-client-registration` work, by pushing abstract json beans which support `quarkus-oidc` `UserInfo` and `TokenIntrospection` down to `quarkus-oidc-common`, but I'm not sure having a shared OIDC client interface works:
 
*  `UserInfo` and `TokenIntrospection` are already part of the `quarkus-oidc` API, moving them down to some other package will impact `quarkus-oidc` users with deprecations
* `quarkus-oidc` client is really focused on authorization code flow, on acquiring verification keys - which is out of scope for `quarkus-oidc-client` - which is meant to work with any token grant type, which does not need an option to get verification keys

So, as far as `quarkus-oidc--client` is concerned, thinking even more about it, I'd just add its own `TokenIntrospection` extending the shared abstract JSON bean and be done with #39298. And may be I'd also do the same for `UserInfo` but for the type of token grants `quarkus-oidc-client` is primarily designed to work with, like `client_credentials` or jwt bearer, `UserInfo` won't be returned. `quarkus-oidc-client` can be used to support custom authorization code flow implementations but it is not something the command line tools Stian has in mind can do it.

So, to summarize, right now, I'm not certain at all a shared interface idea is worth it given that probably the only missing piece `quarkus-oidc-client` is the ability to introspect acquired tokens.

File: integration-tests/oidc-wiremock-logout/src/main/java/io/quarkus/it/keycloak/SecurityEventListener.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import io.quarkus.oidc.AccessTokenCredential;
+import io.quarkus.oidc.RefreshToken;
+import io.quarkus.oidc.SecurityEvent;
+import io.quarkus.oidc.runtime.OidcProviderClient;
+
+@ApplicationScoped
+public class SecurityEventListener {
+
+    public void event(@ObservesAsync SecurityEvent event) {
+        if (SecurityEvent.Type.OIDC_BACKCHANNEL_LOGOUT_COMPLETED == event.getEventType()) {
```
Comment (michalvavrik) [inline integration-tests/oidc-wiremock-logout/src/main/java/io/quarkus/it/keycloak/SecurityEventListener.java:15]: This looks useful.

**Comment by michalvavrik** [comment]
Got it @sberyozkin , I had a look again and I think this is useful. I suppose you will add docs when you make this ready for a review, so just one question:

- until now users didn't really use `OidcProviderClient`, correct? If so, would it make sense to make it an implementation and expose as CDI bean only an interface that won't be in a runtime module? I think that together with javadoc explaining what the interface methods do would make it nicer

**Comment by sberyozkin** [comment]
Hi @michalvavrik 

>until now users didn't really use OidcProviderClient, correct? If so, would it make sense to make it an implementation and expose as CDI bean only an interface that won't be in a runtime module (aka add interface for this class, maybe rename the impl. class)? I think that together with javadoc explaining what the interface methods do would make it nicer

Sure, this is what I was thinking about as well, in the description with `the only remaining thing that I believe should be done is that OidcProviderClient should become an API interface, with the runtime subpackage providing an implementation`...

I've been thinking quite a lot during the last week about how to share interfaces between `quarkus-oidc-client`, `quarkus-oidc`, thanks for suggesting it. For example, I thought, what if we introduce some common interface which offers introspection, userinfo, revocation... But then I was invariably got myself asking, what is the goal, we don't really expect `quarkus-oidc` users do any of these calls, `revocation` is really the only action they may want to do, for example, on the logout events or on some other events, like authentication or authorization failures.

I may be repeating it, but I guess trying to come up with a shared interface to support these 2 extensions whose user audiences and goals are different may not be particularly useful and be rather constraining.

I think what we can definitely try to do is to minimize duplication between `OidcClientImpl` and what will become `OidcProviderClientImpl`, as you have seen, there is some duplication to setup authentication...

As far as what `OidcProviderClient` interface should provider I start leaning to starting with a minimum set of methods that users might actually try to use for some practical purposes, namely, `revocation`, `token introspection` and `userinfo`... I don't think `quarkus-oidc`  users should see JWK key and token acquisition methods... We might add them later if necessary to the interface if we will see some evidence users are really interested...

After this PR is reviewed and approved, I can try to do #39298 by going after introducing shared `TokenIntrospection` and `UserInfo` beans, I can probably get `quarkus-oidc` `TokenIntrospection` and `UserInfo` just extend the common beans.

**Comment by sberyozkin** [comment]
@michalvavrik Hi Michal, so I've introduced `OidcProviderClient` interface. I've probably started at it and `OidcClient` interfaces for an hour again :-). 
My current thinking is as follows:
* `quarkus-oidc-client` `OidcClient` is about supporting a host of token grants to `acquire` and refresh tokens, used in filters and sometimes directly by the users, it is generic in nature. The current endpoint is not the target of these tokens. 
* `quarkus-oidc` `OidcProviderClient` is about helping users to enhance the endpoint code which deals with `verifying` `incoming` bearer or authorization code flow tokens.
* `UserInfo`, `TokenIntrospection` are of primary interest to `quarkus-oidc` users.
* `quarkus-oidc` users are not really expected to deal with `OidcProviderClient` interface directly in most cases, `UserInfo`, `TokenIntrospection` will be typically done for them by `quarkus-oidc` - however this interface will now allow them, for example, to get a delayed UserInfo acquisition, or check the cached token status by introspecting it. Token revocation will most likely be the main case when they may want to 
* Right now, `quarkus-oidc` `OidcProviderClient` does not offer and will most likely not offer, options to get JWK keys to verify, manage the authorization code flow tokens acquisition or refresh - it should really remain under control of `quarklus-oidc`
* Right now, both interfaces have a bit of intersection, token revocation, and might get more intersection, related to the token introspection, userinfo - I'm not too concerned because the use case where `quarkus-oidc-client` `OidcClient` may have to be used to introspect tokens or get UserInfo would most likely be confined to the future CLI tool - these operations won't be used by OIDC client filters - these are are server-side operations. The focus is on helping out users to do fine grained control of logout, token verification failure events to do the revocation for ex.
* I think we may be able to eventually introduce `OidcCommonClient` interface - which will have the current `quarkus-oidc` `OidcProviderClient` methods pushed down to it, and both `OidcProviderClient` and `OidcClient` just extending it. That will end up being transparent to users with respect to the token revocation at least - we may be able to achieve the transparency even for the the userinfo and token introspection beans

What are your thoughts about it ? I plan to add some docs next.

**Comment by sberyozkin** [comment]
I've added initial docs. I'm not really focusing there on `OidcProviderClient` which has JavaDocs, but on the main use case it is supposed to support. Cases where users may want to use it to get `UserInfo` or do the token introspection are advanced.

**Comment by michalvavrik** [comment]
I'll read all tomorrow evening @sberyozkin , quite busy on finishing other PR ATM.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 2]
```diff
@@ -1586,6 +1586,94 @@ public class SecurityEventListener {
 
 TIP: You can listen to other security events as described in the xref:security-customization.adoc#observe-security-events[Observe security events] section of the Security Tips and Tricks guide.
 
+
+[[oidc-token-revocation]]
+=== Token revocation
+
+Sometimes, you may want to revoke the current authorization code flow access and/or refresh tokens.
+For example, if the user has logged out, your logout event listener can revoke access refresh tokens associated with the currenty session.
```
Comment (michalvavrik) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1594]:
```suggestion
For example, if the user has logged out, your logout event listener can revoke access refresh tokens associated with the current session.
```

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 3]
```diff
@@ -1586,6 +1586,94 @@ public class SecurityEventListener {
...
+
+`quarkus.oidc.OidcProviderClient` which provides access to the OIDC provider's UserInfo, token introspection and revocation endpoints, can be used to support such cases.
```
Comment (michalvavrik) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: It reads bit strange to me to start with code `quarkus.oidc.OidcProviderClient`. Not sure if `The` belongs there, but let see if anyone else things that too. If not, it's OK.

**Review by michalvavrik — APPROVED** [review]
Looks very good!

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 3 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: @michalvavrik Yeah, I'm not sure either :-), let me change this sentence slightly and drop the package name as it will be shown in the example below

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc]: @michalvavrik I've kept the package name but restructured a bit, from what I've seen, omitting `the` seems acceptable when referring to (Java) class names, but it can be easily tuned if needed

**Comment by sberyozkin** [comment]
@michalvavrik Thanks for the review. 

I've found it a bit tricky to expose some of these methods cleanly. Currently `quarkus-oidc` has `OidcProvider` and `OidcProviderClient`. The latter is only about sending requests, the former if about keeping keys, having some verification logic.

On `main`, `OidcProviderClient` returns `UserInfoResponse` as opposed to `UserInfo`, because `UserInfo` can be a signed JWT in which case `OidcProvider` verifies it and converts JWT to `UserInfo`. But we can't really expect users do it themselves.

So, in this PR I pushed some of `OidcProvider` code down to `OidcProviderClient` so that no matter what format UserInfo is returned in, the user gets `UserInfo`. This is a bit hacky though, I think eventually, `OidcProvider` and `OidcProviderClient` should become a single `OidcProviderClient`, with the actual verification logic pushed up from `OidcProvider` up to `OidcIdentityProvider` - but this is a sensitive work that requires some dedicated PRs at a later stage :-)

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 3 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc]: Thanks, read well now.

**Comment by sberyozkin** [comment]
@FroMage Have a look please next week if you can at the docs and comments clarifying some design decisions, I believe with this PR you can have an Apple logout and token revocation issue resolved quite neatly.
Michal is happy enough with the PR. 

In general, I'm thinking it will be hard to create a uniform client interface between `quarkus-oidc` and `quarkus-oidc-client` extensions because OIDC one is dealing with verifying incoming tokens and the OIDC client one is a generic interface designed to support many type of grants. `quarkus-oidc` users in most cases do not need to deal with this interface therefore the goals of trying to design some shared interface between two extensions are unclear.

Going forward I may be able to reuse some beans like `UserInfo`, `TokenIntrospection` between both types of clients.

**Review by FroMage — CHANGES_REQUESTED** [review]
Added comments and questions about the API and docs, but I haven't checked the impl code.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 4]
```diff
@@ -1586,6 +1586,94 @@ public class SecurityEventListener {
...
 TIP: You can listen to other security events as described in the xref:security-customization.adoc#observe-security-events[Observe security events] section of the Security Tips and Tricks guide.
 
+
+[[oidc-token-revocation]]
+=== Token revocation
+
+Sometimes, you may want to revoke the current authorization code flow access and/or refresh tokens.
+For example, if the user has logged out, your logout event listener can revoke access refresh tokens associated with the current session.
+
+To support such cases, you can use `quarkus.oidc.OidcProviderClient` which provides access to the OIDC provider's UserInfo, token introspection and revocation endpoints.
```
[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: @FroMage Back channel logout callback from OIDC server to Quarkus will happen if the session is removed (=> ID token), may be access token is also removed but the refresh token may stay alive. 
But it makes sense to update the docs to show an RP initiated logout event (when the user calls a Quarkus logout endpoint which initiates a user redirect to the OIDC logout endpoint), I'll keep the same test code if you don't mind though, it does not really matter which logout event is used in the test

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 5]
```diff
@@ -1586,6 +1586,94 @@ public class SecurityEventListener {
...
+import io.quarkus.oidc.OidcProviderClient;
+import io.quarkus.oidc.RefreshToken;
+import io.quarkus.oidc.SecurityEvent;
+
+@ApplicationScoped
+public class SecurityEventListener {
+
+    public void event(@ObservesAsync SecurityEvent event) {
+        if (SecurityEvent.Type.OIDC_BACKCHANNEL_LOGOUT_COMPLETED == event.getEventType()) { <1>
+            OidcProviderClient oidcProvider = event.getSecurityIdentity().getAttribute(OidcProviderClient.class.getName()); <2>
```
[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1617]: @FroMage The injection does not work in the event observer method, but the injection is shown in the local logout example just below. As far as a new SecurityIdentity method is concerned, this client is not really part of the verified identity state as such, and it is OIDC specific, while SecurityIdentity is not

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 6]
```diff
@@ -1586,6 +1586,94 @@ public class SecurityEventListener {
...
+    @Inject
+    AccessTokenCredential accessToken;
+
+    @Inject
+    RefreshToken refreshToken;
+
+    @GET
+    @Path("logout")
+    public String logout() {
+        oidcSession.logout().await().indefinitely(); <1>
```
[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1664]: @FroMage Sure, I'll just update the example

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: @FroMage Perhaps a better example here would be to show the server revoking tokens in case of the verification failures, before SecurityIdentity is even created, but right now `SecurityEvent` won't give the actual invalid token to revoke, we have an issue to enhance it

[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: @FroMage I've been thinking about it, my understanding it can also depend on how, especially home-grown OIDC providers are implemented, the recent request to support the token revocation came in context of the user logout situation..., so it is quite likely active access tokens are retained sometimes even if the session is gone

[the remark this answers is not in the document]
Comment (FroMage) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: The real reason for invalidating tokens, and this is precisely what Apple documents when it requires that we invalidate tokens, is when users delete their account locally. Presumably, users in this case absolutely do not want a back-channel logout, but they do want their tokens to be revoked.

[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]: @FroMage Sure, I've replaced that with RP initiated logout event, but let me just delete this example from docs altogether, the example with the local `OidcSession` logout will remain and support the main case related to the logout and token revocation. I'll reintroduce the event observer example for revoking invalid tokens, once #45207is resolved

[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc]: @FroMage Yes, it is less confusing now with the the local OidcSession logout only shown in the docs

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 7]
```diff
@@ -1586,6 +1586,56 @@ public class SecurityEventListener {
...
 TIP: You can listen to other security events as described in the xref:security-customization.adoc#observe-security-events[Observe security events] section of the Security Tips and Tricks guide.
 
+
+[[oidc-token-revocation]]
+=== Token revocation
+
+Sometimes, you may want to revoke the current authorization code flow access and/or refresh tokens.
+You can revoke tokens with `quarkus.oidc.OidcProviderClient` which provides access to the OIDC provider's UserInfo, token introspection and revocation endpoints.
+
+For example, when a local logout with <<oidc-session,OidcSession>> is performed, you can use an injected `OidcProviderClient` to remove access and refresh tokens associated with the current session:
```
Comment (FroMage) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:1596]:
```suggestion
For example, when a local logout with <<oidc-session,OidcSession>> is performed, you can use an injected `OidcProviderClient` to revoke access and refresh tokens associated with the current session:
```

**Review by FroMage — APPROVED** [review]
